File: components/frontend/src/changelog/ChangeLog.js
import React, { useEffect, useState } from 'react';
import { Icon } from 'semantic-ui-react';
import { Button, Feed, Form, Header, Segment } from '../semantic_ui_react_wrappers';
import { get_changelog } from '../api/changelog';
import { Avatar } from '../widgets/Avatar';
import { TimeAgoWithDate } from '../widgets/TimeAgoWithDate';
import './ChangeLog.css';

function Event({ description, email, timestamp }) {
    return (
        <Feed.Event>
            <Feed.Label><Avatar email={email} /></Feed.Label>
            <Feed.Content>
                <Feed.Summary>
                    {description}
                    <Feed.Date><TimeAgoWithDate date={timestamp} /></Feed.Date>
                </Feed.Summary>
            </Feed.Content>
        </Feed.Event>
    )
}

function ChangeLogWithoutMemo({ report_uuid, subject_uuid, metric_uuid, source_uuid, timestamp }) {
    const [changes, setChanges] = useState([]);
    const [nrChanges, setNrChanges] = useState(5);
    useEffect(() => {
        let didCancel = false;
        let uuids = {};
        if (report_uuid) { uuids.report_uuid = report_uuid }
        if (subject_uuid) { uuids.subject_uuid = subject_uuid }
        if (metric_uuid) { uuids.metric_uuid = metric_uuid }
        if (source_uuid) { uuids.source_uuid = source_uuid }
        get_changelog(nrChanges, uuids).then(function (json) {
            if (!didCancel) {
                setChanges(json.changelog || []);
            }
        });
        return () => { didCancel = true; };
    }, [report_uuid, subject_uuid, metric_uuid, source_uuid, timestamp, nrChanges]);

    let scope = "Changes in this instance of Quality-time";
    if (report_uuid) { scope = "Changes in this report" }
    if (subject_uuid) { scope = "Changes to this subject" }
    if (metric_uuid) { scope = "Changes to this metric and its sources" }
    if (source_uuid) { scope = "Changes to this source" }

    return (
        <Form>
            <Header size="small">
                {scope}
                <Header.Subheader>Most recent first</Header.Subheader>
            </Header>
            <Segment>
                <Feed size="small">
                    {changes.map(change =>
                        <Event key={change.timestamp + change.delta} description={change.delta} email={change.email} timestamp={change.timestamp} />)}
                </Feed>
                <Button basic icon primary size='small' onClick={() => setNrChanges(nrChanges + 10)}>
                    <Icon name="refresh" /> Load more changes
                </Button>
            </Segment>
        </Form>
    )
}

// Use React.memo so the ChangeLog is not re-rendered on reload, but only when one of its props change
export const ChangeLog = React.memo(props => <ChangeLogWithoutMemo {...props} />)

File: components/server/src/external/routes/changelog.py
"""Changelog routes."""

from typing import cast

import bottle
from pymongo.database import Database

from database import measurements, reports
from server_utilities.type import MetricId, ReportId, SourceId, SubjectId


def _get_changelog(database: Database, nr_changes: str, **uuids: str):
    """Return the recent most nr_changes changes from the changelog."""
    limit = int(nr_changes)
    changes = [
        dict(
            delta=item["delta"]["description"],
            email=item["delta"].get("email", ""),
            timestamp=item.get("timestamp") or item["start"],
        )
        for item in list(measurements.changelog(database, limit, **uuids))
        + list(reports.changelog(database, limit, **uuids))
    ]
    return dict(changelog=sorted(changes, reverse=True, key=lambda change: cast(str, change["timestamp"]))[:limit])


@bottle.get("/api/v3/changelog/source/<source_uuid>/<nr_changes>", authentication_required=False)
def get_source_changelog(source_uuid: SourceId, nr_changes: str, database: Database):
    """Return the recent most nr_changes changes from the source changelog."""
    return _get_changelog(database, nr_changes, source_uuid=source_uuid)


@bottle.get("/api/v3/changelog/metric/<metric_uuid>/<nr_changes>", authentication_required=False)
def get_metric_changelog(metric_uuid: MetricId, nr_changes: str, database: Database):
    """Return the recent most nr_changes changes from the metric changelog."""
    return _get_changelog(database, nr_changes, metric_uuid=metric_uuid)


@bottle.get("/api/v3/changelog/subject/<subject_uuid>/<nr_changes>", authentication_required=False)
def get_subject_changelog(subject_uuid: SubjectId, nr_changes: str, database: Database):
    """Return the recent most nr_changes changes from the subject changelog."""
    return _get_changelog(database, nr_changes, subject_uuid=subject_uuid)


@bottle.get("/api/v3/changelog/report/<report_uuid>/<nr_changes>", authentication_required=False)
def get_report_changelog(report_uuid: ReportId, nr_changes: str, database: Database):
    """Return the recent most nr_changes changes from the report changelog."""
    return _get_changelog(database, nr_changes, report_uuid=report_uuid)


@bottle.get("/api/v3/changelog/<nr_changes>", authentication_required=False)
def get_changelog(nr_changes: str, database: Database):
    """Return the recent most nr_changes changes from the changelog."""
    return _get_changelog(database, nr_changes)
